Lightweight storefront: Pages selector in Themes Preview

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #11471, #11472

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR updates the Themes Preview view with the Pages selector UI:
| Pages selector shown at the top | Page selection sheet shown when tapped | "Preview" text shown if no Pages found |
|-|-|-|
| ![Simulator Screenshot - iPhone 14 Pro - 2023-12-19 at 12 07 25](https://github.com/woocommerce/woocommerce-ios/assets/266376/e59ee45d-82c7-456b-8d70-688f900067cf)  | ![Simulator Screenshot - iPhone 14 Pro - 2023-12-19 at 12 07 30](https://github.com/woocommerce/woocommerce-ios/assets/266376/f6b4916a-86ec-45fa-8a30-a76c510ba721) | ![Simulator Screenshot - iPhone 14 Pro - 2023-12-19 at 12 07 46](https://github.com/woocommerce/woocommerce-ios/assets/266376/b81f6b60-c539-491e-9012-91d32c2ee85c) |

Things that will be handled in a separate PR:

1.  **the design of the page selector area is still subject to change** so it doesn't match the Figma design. The Figma design does not properly match the available space in an iOS toolbar.
2. There is no loading state for when a new page is currently being loaded in the webview, so the transition might look jarring. I attempted it for this PR but it's more complex than I thought (suggestions welcome!)

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Start with a WooExpress site,
3. Go to Menu > Themes
4. Pick a theme
5. If you picked Attar, it should only show "Preview",
6. If you picker another theme, it should show the Page selector like in the design above,
7. Tap the selector, pick a page,
8. Ensure the page is loaded in the webview (check the content of the webpage)
9. Ensure that the selected page title is now shown on the Page selector

## Video

https://github.com/woocommerce/woocommerce-ios/assets/266376/28f8c26d-aefc-43d1-a687-665e1f58d7d0

hafizrahman committed:
Start viewmodel

hafizrahman committed:
Use the view model in the related views.

hafizrahman committed:
Start building loading states.

hafizrahman committed:
Add page selector toolbar UI.

hafizrahman committed:
Add handling for .wordpress.com themes.

hafizrahman committed:
Refactor button placement.

hafizrahman committed:
Add selectedPage and logic to initialize it.

hafizrahman committed:
Set selectedPage's title as selector's title.

hafizrahman committed:
Show sheet with list of items. Also refactors the view model to make it happen.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift

hafizrahman committed:
Build pages list sheet.

hafizrahman committed:
Pass selectedPage's link as the url to be loaded by the current web view.

hafizrahman committed:
Add unit tests.

hafizrahman committed:
Remove "wordpress.com" theme demo case since the theme with such case (Attar) is no longer included.

hafizrahman committed:
Append to pages so that home page is also included.

hafizrahman committed:
Update various colors.

selanthiraiyan reviewed (commented):
Thanks for working on this, @hafizrahman!

I think it will be better if you take a look at the failing unit tests before we review this PR. (In case any code changes are required)

What do you think?

hafizrahman committed:
Add initial assignment for pages

hafizrahman commented:
@selanthiraiyan thanks! I missed a commit. Should be good to go now.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -72,6 +72,10 @@ struct WebView: UIViewRepresentable {
             uiView.reload()
         }
+
+        if uiView.url != url {
+            uiView.load(URLRequest(url: url))
+        }
```
Can you explain this code? 

1. Would this affect other screens outside the Themes Preview screen? (As this isn't checking for `shouldReloadOnUpdate` value)
1. How are `reload` and `load` different? Do we have to do both function calls?

hafizrahman replied:
1. `shouldReloadOnUpdate` is used for the layout switcher's functionality. Layout switching depends on re-applying Javascript on the same URL, and the re-applying is done by reloading that same URL.
2. The `uiView.load()` call is used for the page switcher functionality. To load a new page, the parent View passes the new URL to this `WebView`. In this situation, the UIView is not re-created with the new URL, so instead  `updateUIView()` have to be used to ask the actual `WKWebView` to load that new page. 

> (As this isn't checking for shouldReloadOnUpdate value)

Can you explain what you meant with this?

hafizrahman replied:
I added a comment in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/297d47791683669dcdb05e6234bda10fdc670b87 to explain point 2 above.

selanthiraiyan replied:
> Can you explain what you meant with this?

- `uiView.load(URLRequest(url: url))` will execute even if `shouldReloadOnUpdate` is false. 
- But, if we take `uiView.reload()` method call, checking `shouldReloadOnUpdate` before calling the method ensures that this change won't affect other screens outside Themes. (As currently `shouldReloadOnUpdate` value is set as `true` only when used in `ThemesPreviewView`)

The `uiView.load(URLRequest(url: url))` will be executed for all other screens where `WebView` is used. We need to ensure that this change won't have any undesired effects. 

What do you think? 

1. Should the `uiView.load(URLRequest(url: url))` call be executed for all screens using `WebView`? 
1. What happens when we call `uiView.reload()` first and then `uiView.load(URLRequest(url: url))` for the themes screen case? Is it necessary to call both methods? 

Tagging @itsmeichigo to get their opinion on this.

hafizrahman replied:
> Should the uiView.load(URLRequest(url: url)) call be executed for all screens using WebView?

Yes, if the parent View in other screens need to programmatically update the URL in the `WebView` call (like what's happening here when changing Page manually), they want this to be called so that the new URL is actually loaded.

I wonder if `shouldReloadOnUpdate` can be renamed to something else to make it clearer what it does. Here is the logic with the layout change:

1. When a new layout is selected in the menu, then `selectedDevice` is updated,
2. This in turns redraws the `WebView` since the `onCommit` parameter is updated:
```
                    WebView(
                        isPresented: .constant(true),
                        url: url,
                        shouldReloadOnUpdate: true,
                        disableLinkClicking: true,
                        onCommit: { webView in
                            webView.evaluateJavaScript(self.selectedDevice.viewportScript)
                        }
                    )
```
3. For the new JavaScript in `onCommit` to be called, the page also needs to be reloaded, since the JS call depends on the `didCommit` event below to happen.

```
        func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
            parent.onCommit?(webView)
        }
```
4. By default this is not happening, though, since the `url` passed in the `WebView` call remains the same, thus the webview won't reload it.
5. Thus, we need to also force the reload somehow. This is where the `shouldReloadOnUpdate` parameter comes in. If this is true, then whenever there's any `WebView` redraw/update, then it should reload the screen. This in turn calls the JavaScript again, which is needed to change the layout.

> What happens when we call uiView.reload() first and then uiView.load(URLRequest(url: url)) for the themes screen case? Is it necessary to call both methods?

I think I see what you mean, though. I suppose it should be like this instead?
```
        if shouldReloadOnUpdate && uiView.url == url {
            uiView.reload()
        }
```

itsmeichigo replied:
I think it makes sense to load the request on state change. I found [an article](https://sarunw.com/posts/swiftui-webview/) that advises loading requests even without checking for URL changes. I experimented with this and found that reloading is not necessary if we do that:

```Swift
    func updateUIView(_ uiView: WKWebView, context: Context) {
        uiView.load(URLRequest(url: url))
    }
```

The reason is that the web view is rendered again when the preview view's `selectedDevice` changes, which triggers `updateUIView`. The web view is then asked to load the given URL again and the correct layout is rendered with the injected JS script. If we follow this, the load request in `makeUIView` is redundant. We can also get rid of `shouldReloadOnUpdate`.

selanthiraiyan replied:
Nice. Thanks Huong! 🙇 

@hafizrahman  Let us follow Huong's suggestion and get rid of `shouldReloadOnUpdate`.

hafizrahman replied:
Nice info in that article. Makes me wonder why `fun updateUIView()` was empty before, perhaps because the `WebView` is not expected to handle state change. I'll update & simplify the code.

hafizrahman replied:
Added in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/91c59aa7db17b438bab2a5357d871977bd1e027e

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -126,6 +140,56 @@ struct ThemesPreviewView: View {
+        }
+    }
+
+    private func pageSelector(state: ThemesPreviewViewModel.State, pageTitle: String) -> some View {
+        switch state {
+        case .pagesLoading:
+            return AnyView(Text(Localization.preview))
+
+        case .pagesContent:
+            return AnyView(
+                Button(action: { showPagesMenu = true },
+                       label: {
+                           HStack {
+                               Text(pageTitle).bodyStyle()
+                               Image(uiImage: .chevronDownImage)
+                                   .fixedSize()
+                                   .secondaryBodyStyle()
+
+                           }
+                       })
+            )
+
+        case .pagesLoadingError:
+            return AnyView(Text(Localization.preview))
+        }
+    }
```
Please rewrite this using `ViewBuilder` to avoid `AnyView` usage. [Why?](https://www.swiftbysundell.com/articles/avoiding-anyview-in-swiftui/) You can directly access the viewModel instead of passing parameters.

hafizrahman replied:
Great tips, thanks! Added in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/98fd07bf35f6c4bf9e9dfb75895a462fcd5b47c6

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -126,6 +140,56 @@ struct ThemesPreviewView: View {
+                               Text(pageTitle).bodyStyle()
+                               Image(uiImage: .chevronDownImage)
+                                   .fixedSize()
+                                   .secondaryBodyStyle()
```
The theme picker dropdown downfacing arrow should be black color. The current color looks gray. 
<img width="155" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/7c68e51c-761e-497b-aee3-31e91ddbd18e">

hafizrahman replied:
Good eye! https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/754cd51f32bbadee86cb935d996a2dd98cc8702b

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -193,12 +264,19 @@ private extension ThemesPreviewView {
             comment: "The error message shown if the app can't show a theme demo."
         )
+
+        static let pagesSheetHeading = NSLocalizedString(
+        "ThemesPreviewView.pagesSheetHeading",
+        value: "View other store pages on this theme",
+        comment: "Heading for sheet displaying list of pages"
+        )
```
nit: Code indentation needs to be updated.

hafizrahman replied:
right! Added in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/16671b37bad6fb0a0c6a544add412a6d0739eb26

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -193,12 +264,19 @@ private extension ThemesPreviewView {
         )
+
+        static let pagesSheetHeading = NSLocalizedString(
+        "ThemesPreviewView.pagesSheetHeading",
```
```suggestion
        "themesPreviewView.pagesSheetHeading",
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+/// ViewModel for ThemesPreviewView
+///
+final class ThemesPreviewViewModel: ObservableObject {
+    @Published private(set) var pages: [WordPressPage] = []
```
```suggestion
    @Published private(set) var pages: [WordPressPage]
```

We can skip setting an initial value as we are setting this in `init`.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+    @MainActor
+    func fetchPages() async {
+        do {
+            // Append the list of pages to the existing home page value, since the API call result
+            // do not include the home page.
```
Thanks for adding comments.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Themes/ThemesPreviewViewModelTests.swift:
```diff
@@ -0,0 +1,65 @@
+
+        // Then
+        XCTAssertEqual(viewModel.state, .pagesLoading)
+        XCTAssertEqual(viewModel.pages.count, 1) // The default "Home" page.
```
1. You can test the actual contents of the `pages` array to ensure the value is "Home" page. (Instead of checking  the `count`)
1. I recommend moving this "Home" page check to a separate unit test. It is better if one test checks for one thing. It helps to identify issues when a test fails.

hafizrahman replied:
Makes a lot of sense, handled in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/f0c1f448ffa291b920a6f2e25125567062af333f

hafizrahman replied:
I kept the `count` checking because I think it's useful as a check as well.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Themes/ThemesPreviewViewModelTests.swift:
```diff
@@ -0,0 +1,65 @@
+        XCTAssertEqual(viewModel.state, .pagesContent)
+    }
+
+    func test_setSelectedPage() {
```
```suggestion
    func test_setSelectedPage_updates_selectedPage() {
```

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Themes/ThemesPreviewViewModelTests.swift:
```diff
@@ -0,0 +1,65 @@
+        await viewModel.fetchPages()
+
+        // Then
+        XCTAssertEqual(viewModel.pages.count, 4) // 3 pages + 1 default "Home" page.
```
As I mentioned in the previous comment, you can test the actual contents of the pages array. (Instead of checking the count) This will also help test the order of the array.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/f0c1f448ffa291b920a6f2e25125567062af333f

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -63,21 +64,29 @@ struct ThemesPreviewView: View {
+
     private let theme: WordPressTheme
 
     /// Triggered when "Start with this theme" button is tapped.
     var onStart: () -> Void
```
Not related to this PR: 

"Start with This Theme" button text should be "Use This Theme" when in Theme settings page. 
<image src="https://github.com/woocommerce/woocommerce-ios/assets/524475/f9882073-04ec-48fb-808c-41f6cd856227" width="320"/>

1. The button text needs to be fixed.
2. Using a common name for this callback might be better.

selanthiraiyan replied:
I will handle the button name fix while working on this https://github.com/woocommerce/woocommerce-ios/issues/11494 task.

hafizrahman replied:
name update in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/5657fb6594f6ba8eef11f0a8aa1e3b9e27ba53f3

selanthiraiyan reviewed (commented):
This works in general. 🎉 

I left a few comments and questions. 

---

The current theme value is updated as soon as I select a theme to see Preview. We should instead wait for the user to apply the theme before changing current theme value. Logged GH issue https://github.com/woocommerce/woocommerce-ios/issues/11493

https://github.com/woocommerce/woocommerce-ios/assets/524475/30348f16-7d29-4101-bd86-f74e0ff3d707

selanthiraiyan committed:
Merge branch 'trunk' into feat/themes-preview-page-selector

hafizrahman committed:
Fix typo.

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hafizrahman committed:
Improve function name

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hafizrahman committed:
Add comment explaining the need for manual loading.

hafizrahman committed:
Refactor to use @ViewBuilder instead.

hafizrahman committed:
Update chevron color.

hafizrahman committed:
fix indentation

hafizrahman committed:
Refactor unit tests for successful fetching.

hafizrahman committed:
Make function name for main CTA be more generic.

hafizrahman committed:
Remove the need to initialize

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hafizrahman commented:
Thanks for the thorough review @selanthiraiyan . I've addressed all feedbacks, good for another round!

hafizrahman committed:
Add  unit test for failure state.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -126,6 +141,53 @@ struct ThemesPreviewView: View {
+        }
+    }
+
+    private func pagesListSheet(pages: [WordPressPage]) -> some View {
+        return VStack {
+            Text(Localization.pagesSheetHeading)
+                .subheadlineStyle()
+                .padding(Layout.pagesSheetPadding)
+            ForEach(pages) { page in
+                Button(action: {
+                    viewModel.setSelectedPage(page: page)
+                    showPagesMenu = false
+                }, label: {
+                    Text(page.title)
+                        .bodyStyle()
+                        .frame(maxWidth: .infinity, alignment: .leading)
+
+                })
+                .padding(Layout.contentPadding)
+            }
+
+            Spacer()
+        }
```
Opinion:

1. Make this a `var` to keep it consistent with `pageSelector` above. 
1. Remove the `return` statement to improve readability.
1. Remove parameter and read from view model directly. 

```suggestion
    private var pagesListSheet: some View {
        VStack {
            Text(Localization.pagesSheetHeading)
                .subheadlineStyle()
                .padding(Layout.pagesSheetPadding)
            ForEach(viewModel.pages) { page in
                Button(action: {
                    viewModel.setSelectedPage(page: page)
                    showPagesMenu = false
                }, label: {
                    Text(page.title)
                        .bodyStyle()
                        .frame(maxWidth: .infinity, alignment: .leading)
                })
                .padding(Layout.contentPadding)
            }

            Spacer()
        }
```

hafizrahman replied:
Thanks! this makes sense.

selanthiraiyan reviewed (approved):
Thanks for the update, @hafizrahman! This looks good now. 

I am approving to unblock other PRs. This should be good to merge once my comment about the `WebView` change is addressed.

selanthiraiyan commented:
@hafizrahman For future PRs and branches, please mention the GitHub issue number in both the branch name and the PR description for easier tracking. 🙏

hafizrahman committed:
refactor pagesListSheet as var instead.

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hafizrahman committed:
update pagesListSheet usage.

hafizrahman committed:
Shortening the lines.

hafizrahman commented:
>  For future PRs and branches, please mention the GitHub issue number in both the branch name and the PR description for easier tracking. 

Thanks for the reminder! I usually add it in PR description but I guess it got deleted accidentally here. I'll remember to add it in branch name too!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -126,6 +141,50 @@ struct ThemesPreviewView: View {
             .navigationBarTitleDisplayMode(.inline)
         }
+        .task { await viewModel.fetchPages() }
+        .sheet(isPresented: $showPagesMenu) { pagesListSheet }
```
We should set the `presentationDetent` if iOS 16 or above is available: 

```suggestion
        .sheet(isPresented: $showPagesMenu) {
            if #available(iOS 16, *) {
                pagesListSheet
                    .presentationDetents([.medium, .large])
            } else {
                pagesListSheet
            }
        }
```

hafizrahman committed:
Support presentationDetents for iOS 16.

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Update pages list so that its content is not being cut off in medium .presentationDetents

hafizrahman committed:
Simplify WebView behavior by always making it load the url whenever there's a state change.

This then handles both cases (layout change with same URL, and page change with different URL) nicely. It's also recommended by this article https://sarunw.com/posts/swiftui-webview/ to do updateUIView() this way.

hafizrahman commented:
Nice discussion about the WebView behavior, thanks folks. I'm setting this to auto-merge now that I've handled the WebView load behavior.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Themes/ThemesPreviewView.swift:
```diff
@@ -126,6 +140,61 @@ struct ThemesPreviewView: View {
+                    }
+                    Spacer()
+                }
+                .frame(minHeight: geometry.size.height)
```
I am curious. Do we need to set the minHeight using `GeometryReader`? What happens if we don't set it?

hafizrahman replied:
The `.presentationDetents` shows the sheet at a predefined height (medium and large).

When the Pages list is very long, at medium sheet height, the list is somehow shown vertically centered and the top label are cut-off. This is a solution for that.

Video of the issue that this commit solves:


https://github.com/woocommerce/woocommerce-ios/assets/266376/7466d813-1664-43fa-831e-f8658813ddd5

selanthiraiyan replied:
Thanks for the explanation, Hafiz! 

I removed the GeometryReader, and I only kept the ScrollView

```swift
    private var pagesListSheet: some View {
        ScrollView {
            VStack {
                Text(Localization.pagesSheetHeading)
                    .subheadlineStyle()
                    .padding(Layout.pagesSheetPadding)
                ForEach(viewModel.pages) { page in
                    Button(action: {
                        viewModel.setSelectedPage(page: page)
                        showPagesMenu = false
                    }, label: {
                        Text(page.title)
                            .bodyStyle()
                            .frame(maxWidth: .infinity, alignment: .leading)
                    })
                    .padding(Layout.contentPadding)
                }
                Spacer()
            }
        }
    }
```

I don't see this issue in iOS 16.4 simulator. Can you let me know what I am missing here? 

https://github.com/woocommerce/woocommerce-ios/assets/524475/c511d413-6586-44a6-9af4-e6eea5c18798

@itsmeichigo Is using `GeometryReader` required?

hafizrahman replied:
Conversely, when the Pages list is too short, this stretching issue happened, and it's fixed also by this commit:

https://github.com/woocommerce/woocommerce-ios/assets/266376/4412340b-7305-4da0-a244-a38b0091b950

hafizrahman replied:
Huh, weird, I wonder why this happened in my Simulator. It is also in 16.4.

selanthiraiyan replied:
> Conversely, when the Pages list is too short, this stretching issue happened, and it's fixed also by this commit:

I assume that you mean the code without `ScrollView`. Without the scroll view the `VStack` is trying to occupy all the space which is expected behaviour. 

BTW, my question https://github.com/woocommerce/woocommerce-ios/pull/11492#discussion_r1432362813 is about the use of `GeometryReader` and the `.frame(minHeight: geometry.size.height)` modifier.

hafizrahman replied:
Oh I think I see what you meant. ScrollView alone is enough? Testing.

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/11492/commits/9075f71c20b4a80147bd700e141159019069a0d1

selanthiraiyan replied:
Thank you!

hafizrahman committed:
Simplify solution without needing GeometryReader.
